---
title: "R Notebook"
output: html_notebook
---

```{r Load data}
library(WGCNA)
library(DESeq2)
require(RnaSeqAnalysis)
library(tidyverse)
library(plotly)

```

```{r}


load("../ReportOutput/DifferentialOutput.RData")
```

```{r}
# Setup WGCNA data
inputData <- as.data.frame(assay(vsd)) %>%
  filter(rowMeans(counts(dds, normalized = TRUE)) > 20) %>%
  top_n(6000, apply(., 1, var)) %>%
  top_n(5590, -apply(., 1, var)) %>%
  # Rownames get removed when using the piping operator (Hadley hates rownames!)
  mutate(GeneName = row.names(.))

# Re-do rownames 
rownames(inputData) <- inputData$GeneName

# Make data formatted for WGCNA
datExpr <- t(inputData[, -ncol(inputData)])
colnames(datExpr) <- rownames(inputData)

datTraits <- colData(vsd)
### NOTE: DO NOT FILTER BY DIFFERENTIALLY EXPRESSED GENES
### https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/faq.html
# datExpr <- t(counts(dds, normalized = TRUE)[DiffAtleastOnce$SignifOnce,])

# Enable multi-threading if possible
allowWGCNAThreads(nThreads = 6)
```

# Determine soft thresholding

```{r Determine soft-threshold for subsets, cache = TRUE}
contrast <- c("Diet.x", "HFD","chow")
SubsetsFromContrast <- function(colData, contrast) {
  group1 <- which(colData[, contrast[1]] == contrast[2])
  group2 <- which(colData[, contrast[1]] == contrast[3])

  return(setNames(list(group1, group2), nm = contrast[2:3]))
}
subsets <- SubsetsFromContrast(datTraits, contrast)

thresholds.list <- WGCNA_ThresholdSubsets(datExpr, subsets = subsets)

WGCNA_PlotThresholds(thresholds.list$fitIndices)
```

```{r}
save(thresholds.list, file = "SoftThresholding.RData")
```

```{r}
load("SoftThresholding.RData")
```

```{r Calculate blockwiseModules, eval = FALSE}
chosen.power <- max(unlist(thresholds.list$powerEstimates))
NetworkList <- WGCNA_SubsetModules(datExpr[, 1:100], 
                                   subsets = list(HFD = group1, Chow = group2), 
                                   include.total = TRUE,
                                   power = chosen.power)

save(NetworkList, file = "NetworkList.Rdata")
```

```{r eval = FALSE}
# Consensus modules
multiExpr <- list(list(data = datExpr[group1, ]),
                  list(data = datExpr[group2, ]))
NetworkConsensus <- blockwiseConsensusModules(multiExpr, power = chosen.power, maxBlockSize = ncol(datExpr))
```

```{r}
load(file = "NetworkList.Rdata")
TOMList <- list()
for (netname in names(NetworkList)) {
  load(file = NetworkList[[netname]]$TOMFiles)
  TOMList[[netname]] <- TOM
}
rm(TOM)
```

```{r Chow Net dendrogram}
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(NetworkList$Chow$dendrograms[[1]],
                    colors = data.frame(HFD = NetworkList$HFD$mergedColors[NetworkList$Chow$blockGenes[[1]]],
                                       Chow = NetworkList$Chow$mergedColors[NetworkList$Chow$blockGenes[[1]]]),
                    main = "Clusters for Chow",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```

```{r HFD colored by Chow}
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(NetworkList$HFD$dendrograms[[1]],
                    colors = data.frame(HFD = NetworkList$HFD$mergedColors[NetworkList$HFD$blockGenes[[1]]],
                                        Chow = NetworkList$Chow$mergedColors[NetworkList$HFD$blockGenes[[1]]]),
                    main = "Clusters for HFD",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)
```

```{r}
datME <- cbind(data.frame(datTraits)[subsets$HFD,] %>% select(Diet.x, TimeNum, Lesion.area),
               NetworkList$HFD$MEs)
NS1=networkScreening(y=datTraits[subsets$HFD,]$Lesion.area,
                     datME=NetworkList$HFD$MEs,
                     datExpr=datExpr[subsets$HFD,], oddPower=3, 
                     blockSize=1000, minimumSampleSize=4, addMEy=TRUE,
                     removeDiag=FALSE, weightESy=0.5)
GS1 <- cor(datTraits$Lesion.area, datExpr, use = "p")

#network screening analysis 
# mean(NoiseGeneIndicator[rank(NS1$p.Weighted,ties.method="first")<=100]) # standard analysis based on the correlation p-values (or Student T test) 
# mean(NoiseGeneIndicator[rank(NS1$p.Standard,ties.method="first")<=100])

datKME=signedKME(datExpr[subsets$HFD,], NetworkList$HFD$MEs, outputColumnName="MM.") # Display the first few rows of the data frame head(datKME)
```

```{r}
y <- datTraits$Lesion.area[subsets$HFD]
datME <- moduleEigengenes(datExpr[subsets$HFD,], NetworkList$HFD$moduleColors)$eigengenes


# Create a data frame holding the results of gene screening
GeneResultsNetworkScreening=data.frame(GeneName=row.names(NS1), NS1) 
# Write the data frame into a file 
write.table(GeneResultsNetworkScreening, file="GeneResultsNetworkScreening.csv", row.names=F,sep=",") 
# Output of eigengene information: 
datMEy = data.frame(y, datME) 
eigengeneSignificance = cor(datMEy, y);
eigengeneSignificance[1,1] = (1+max(eigengeneSignificance[-1, 1]))/2 
eigengeneSignificance.pvalue = corPvalueStudent(eigengeneSignificance, nSamples = length(y)) 
namesME=names(datMEy) # Form a summary data frame
10
out1=data.frame(t(data.frame(eigengeneSignificance, eigengeneSignificance.pvalue, namesME, t(datMEy)))) 
# Set appropriate row names
dimnames(out1)[[1]][1]="EigengeneSignificance"
dimnames(out1)[[1]][2]="EigengeneSignificancePvalue" 
dimnames(out1)[[1]][3]="ModuleEigengeneName" 
dimnames(out1)[[1]][-c(1:3)]=dimnames(datExpr[subsets$HFD,])[[1]] 
# Write the data frame into a file 
write.table(out1, file="MEResultsNetworkScreening.csv", row.names=TRUE, col.names = TRUE, sep=",") 
# Display the first few rows: 
head(out1)
```

```{r Table of overlapping clusters, eval = FALSE}
table("HFD" = NetworkList$HFD$mergedColors, "Chow" = NetworkList$Chow$mergedColors)
# caret::confusionMatrix(table("HFD" = NetworkList$HFD$mergedColors, "Chow" = NetworkList$Chow$mergedColors))
```

# Multidimensional scaling

## HFD

```{r cache = TRUE}
# cmd1=cmdscale(as.dist(TOM_HFD),3)
# save(cmd1, file = "MDS_Scaling_HFD.RData")
load("MDS_Scaling_HFD.RData")


colorh1 = NetworkList$HFD$moduleColors

pca.plotdata <- data.frame(cmd1, color = as.character(colorh1)) %>%
  filter(color != "grey")

pal <- setNames(as.character(unique(pca.plotdata$color)), unique(pca.plotdata$color))

plot_ly(pca.plotdata, x = ~X1, y = ~X2, z = ~X3) %>%
  add_markers(color = ~color, colors = pal)
```

## Chow

```{r cache = TRUE}
# cmd2=cmdscale(as.dist(TOM_Chow), 3)
# save(cmd2, file = "MDS_Scaling_Chow.RData")
load("MDS_Scaling_Chow.RData")


colorh1 = NetworkList$Chow$moduleColors

pca.plotdata <- data.frame(cmd2, color = as.character(colorh1)) %>%
  filter(color != "grey")

pal <- setNames(as.character(unique(pca.plotdata$color)), unique(pca.plotdata$color))

plot_ly(pca.plotdata, x = ~X1, y = ~X2, z = ~X3) %>%
  add_markers(color = ~color, colors = pal)
```

# 

```{r eval = TRUE, fig.height=10, cache = TRUE}
# Gene connectivity
group <- ifelse(dds$Diet.x == "chow", 0, 1)

ADJ1=abs(cor(datExpr,use="p"))^chosen.power

GeneSignificance = abs(cor(group, datExpr, use = "p"))

# Next module significance is defined as average gene significance. 
ModuleSignificance=tapply(GeneSignificance, NetworkList$HFD$moduleColors, mean, na.rm=T)

Alldegrees1=intramodularConnectivity(ADJ1, NetworkList$HFD$moduleColors)
# datSamples <- colData %>% cbind(., MEs)

# Gene connectivity versus intramodular connectivity
colorlevels=unique(NetworkList$HFD$moduleColors)

## Check PCA scores of eigengens
MEdata = moduleEigengenes(datExpr[subsets$HFD,], NetworkList$HFD$moduleColors, softPower = 16)
varExpl = t(MEdata$varExplained)
rownames(varExpl) = colnames(MEdata$eigengenes)
# write.table(varExpl, file = paste(myOutPath, "eigengene.PC1.scores.txt", sep=""), sep = "\t", quote = F, col.names=F)
minVarExpl = 0.5
validMods = gsub("^ME", "", rownames(varExpl)[varExpl >= minVarExpl])
colorlevels <- validMods

par(mfrow=c(as.integer(0.5+length(colorlevels)/2), 2) )
par(mar = c(4,5,3,1)) 

for (i in c(1:length(colorlevels))) {
  whichmodule=colorlevels[[i]]; 
  restrict1 = (NetworkList$HFD$moduleColors==whichmodule); 
  verboseScatterplot(Alldegrees1$kWithin[restrict1], 
                     GeneSignificance[restrict1], 
                     col=NetworkList$HFD$moduleColors[restrict1], 
                     main=whichmodule,
                     xlab = "Connectivity", ylab = "Gene Significance",
                     abline = TRUE) 
  }
```

```{r Correlate with lesion area, eval = FALSE}
# Gene connectivity
group <- dds$Lesion.area

# ADJ1=abs(cor(datExpr,use="p"))^chosen.power

GeneSignificance = abs(cor(group, datExpr, use = "p"))

# Next module significance is defined as average gene significance. 
ModuleSignificance=tapply(GeneSignificance, NetworkList$HFD$moduleColors, mean, na.rm=T)
Alldegrees1=intramodularConnectivity(ADJ1, NetworkList$HFD$moduleColors)

# datSamples <- colData %>% cbind(., MEs)

# Gene connectivity versus intramodular connectivity
colorlevels=unique(NetworkList$HFD$moduleColors)

## Check PCA scores of eigengens
MEdata = moduleEigengenes(datExpr, NetworkList$HFD$moduleColors, softPower = chosen.power)
varExpl = t(MEdata$varExplained)
rownames(varExpl) = colnames(MEdata$eigengenes)
# write.table(varExpl, file = paste(myOutPath, "eigengene.PC1.scores.txt", sep=""), sep = "\t", quote = F, col.names=F)
minVarExpl = 0.5
validMods = rownames(varExpl)[varExpl >= minVarExpl]
colorlevels <- gsub("^ME", "", validMods)
#


par(mfrow=c(as.integer(0.5+length(colorlevels)/2), 2) )
par(mar = c(4,5,3,1))

for (i in c(1:length(colorlevels))) {
  whichmodule=colorlevels[[i]]; 
  restrict1 = (NetworkList$HFD$moduleColors==whichmodule); 
  verboseScatterplot(Alldegrees1$kWithin[restrict1], 
                     GeneSignificance[restrict1], 
                     col=NetworkList$HFD$moduleColors[restrict1], 
                     main=whichmodule,
                     xlab = "Connectivity", ylab = "Gene Significance",
                     abline = TRUE) 
  }
```

```{r Add eigengenes to sample data}
METraits_HFD <- cbind(datTraits, MEdata$eigengenes)

as.data.frame(METraits_HFD) %>% tidyr::gather(Module, Eigengene, MEblack:MEyellow) %>% 
  ggplot(aes(x = Lesion.area, y = Eigengene, color = Diet.x)) + 
  geom_point() +
  facet_grid(Module~TimeNum, scales = "free_y") + 
  geom_smooth(method = "lm")
```


```{r Differential connectivity, eval = FALSE}
# Test connectivity
contrast <- c("Diet.x", "HFD", "chow")

#
n.perm <- 2
# 
DiffConn <- DifferentialConnectivity(datExpr, colData(vsd), contrast, softThreshold = chosen.power)
DiffRegions <- DiffConn %>% mutate(RegionCounts = map2(tstat, DiffK, CountRegions),
                                   Region = unlist(map(RegionCounts, function(x) ifelse(any(x == 1),
                                                                                        which(x == 1),
                                                                                        0))),
                                   Module = NetworkList$Total$moduleLabels, Color = NetworkList$Total$moduleColors)

palHFD <- setNames(as.character(unique(NetworkList$HFD$moduleColors)), unique(NetworkList$HFD$moduleColors))
palChow <- setNames(as.character(unique(NetworkList$Chow$moduleColors)), unique(NetworkList$Chow$moduleColors))

ggplotly(DiffRegions %>% 
           ggplot(aes(x = DiffK, y = tstat, color = NetworkList$HFD$moduleColors)) + 
           geom_point(alpha = 0.8) +
           geom_hline(yintercept = 1.96, col = "red") + geom_hline(yintercept = -1.96, col = "red") +
           scale_color_manual(values = palChow) +
           geom_vline(xintercept = 0.4, col = "red") + geom_vline(xintercept = -0.4, col = "red") +
           ggtitle(paste("Colored by Chow Modules, cor=",
                         signif(cor(DiffRegions$DiffK,DiffRegions$tstat,use="pairwise.complete.obs"),3))))


PermConn <- PermuteConnectivity(datExpr, as.data.frame(colData(dds)), contrast, n.perm = n.perm, softThreshold = chosen.power)
save(PermConn, file = "PermutationConnectivity.Rdata")
load("PermutationConnectivity.RData")

PermSumm <- PermConn %>% group_by(Rep) %>% summarise(RegionCounts = list(CountRegions(tstat, DiffK))) %>%
  unnest() %>% mutate(Region = rep(1:8, max(.$Rep)))

DiffRegionsums <- CountRegions(DiffConn$tstat, DiffConn$DiffK)
LogicalSums <- PermSumm %>% ungroup() %>% group_by(Region) %>%
  summarise(LogicalSum = sum(RegionCounts < DiffRegionsums[[mean(Region)]]),
            TotalPerm = n()) %>%
  mutate(Pval = (LogicalSum + 1) / (TotalPerm + 1))
  

# DiffRegions <- TestRegions(DiffConn$tstat, DiffConn$DiffK)
```

```{r correlations, eval = FALSE}
source("WGCNA_HelperFuns.R")
myOutPath <- "CorrelationPlots/"
select = dplyr::select

if (!dir.exists(myOutPath)) dir.create(myOutPath)

## Correlate module eigengenes with disease endpoints
datTraitsHFD = as.data.frame(datTraits) %>% filter(Diet.x == "HFD")
datTraitsChow = as.data.frame(datTraits) %>% filter(Diet.x == "chow")
datTraits <- as.data.frame(datTraitsHFD) %>% select(TimeNum, Lesion.area:RatioMildSevere)

## Check PCA scores of eigengens
MEdata = moduleEigengenes(datExpr, NetworkList$HFD$moduleColors, softPower = chosen.power)
varExpl = t(MEdata$varExplained)
rownames(varExpl) = colnames(MEdata$eigengenes)
# write.table(varExpl, file = paste(myOutPath, "eigengene.PC1.scores.txt", sep=""), sep = "\t", quote = F, col.names=F)
minVarExpl = 0.5
validMods = rownames(varExpl)[varExpl >= minVarExpl]



nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, NetworkList$HFD$moduleColors)$eigengenes
MEs = orderMEs(MEs0[,validMods])
moduleTraitCor = cor(MEs, datTraitsHFD, use = "p")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)

moduleTraitCor.sp = moduleTraitCor
moduleTraitPvalue.sp = moduleTraitPvalue

for(mod in colnames(MEs)) {
    for(trait in colnames(datTraits)) {
        r = cor.test(MEs[,mod], datTraits[,trait], method="spearman")
        moduleTraitCor.sp[mod, trait] = r$estimate
        moduleTraitPvalue.sp[mod, trait] = r$p.value
    }
}
```